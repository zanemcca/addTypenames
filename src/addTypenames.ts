import * as _reduce from 'lodash/reduce';
import * as _map from 'lodash/map';
import * as _keys from 'lodash/keys';
import * as _filter from 'lodash/filter';
import {
    GraphQLSchema,
    GraphQLType,
    getNamedType,
    GraphQLNamedType,
    GraphQLField,
    isUnionType,
    isInterfaceType,
    GraphQLObjectType,
    isObjectType,
    isAbstractType,
    isInputObjectType,
    isScalarType,
    GraphQLScalarType,
} from 'graphql';

interface AddTypenamesProps {
    item: any;
    schema: GraphQLSchema;
    propertyOfParent?: string;
    parentTypeName?: string;
}

const isPromise = (possiblePromise: any): possiblePromise is Promise<unknown> =>
    typeof possiblePromise === 'object' && possiblePromise.then && typeof possiblePromise.then === 'function';

export const getCandidateTypes = ({ item, schema }: AddTypenamesProps) => {
    if (Array.isArray(item)) {
        throw new Error(`getCandidateTypes expects objects to be passed in not arrays!`);
    }
    if (!item || typeof item !== 'object' || item instanceof Date) {
        return [];
    }
    const typeMap = schema.getTypeMap();
    const fields = Object.keys(item).sort();
    const candidateTypes: GraphQLType[] = _filter(typeMap, possibleType => {
        if (isInputObjectType(possibleType)) {
            return false;
        }
        const config = possibleType.toConfig() as any;
        if (!config.astNode) {
            // Internal types do not have astNode set
            return false;
        }
        const doAllGivenFieldsExistInThePossibleType = config.fields && !fields.some(field => !config.fields[field]);
        return doAllGivenFieldsExistInThePossibleType;
    });

    return candidateTypes;
};

export const addTypenames = ({
    item,
    schema,
    propertyOfParent,
    parentTypeName,
}: AddTypenamesProps): Promise<any> => {
    if (Array.isArray(item)) {
        throw new Error(`addTypenames expects objects to be passed in not arrays!`);
    }
    if (item === null || item instanceof Date || typeof item !== 'object') {
        return item;
    }

    const newObject = { ...item };
    let namedType: GraphQLNamedType;

    if (item.__typename) {
        namedType = schema.getType(item.__typename)!;
    } else {
        if (parentTypeName) {
            if (!propertyOfParent) {
                throw new Error(`Expected a propertyOfParent to be given when parentTypeName is given!`);
            }
            const parentType = schema.getType(parentTypeName);
            if (!parentType) {
                throw new Error(`No parent type found for the typename ${parentTypeName}`);
            }

            if (!isObjectType(parentType)) {
                throw new Error(`Parent type ${parentTypeName} is expected to be an object type but it not`);
            }

            const field = parentType.getFields()[propertyOfParent];
            if (!field) {
                throw new Error(`The given field (${propertyOfParent}) does not exist on type ${parentType.name}`);
            }
            namedType = getNamedType(field.type);
        } else if (propertyOfParent) {
            throw new Error(`Expected a parentTypeName to be given when propertyOfParent is given!`);
        } else {
            const errorHint = `

This is the relevant object given\n${JSON.stringify(item, null, '\t')}

Hint: Try adding a __typename on your initial input object or pass in a propertyOfParent & parentTypeName.
`;
            const candidateTypes = getCandidateTypes({ item, schema });
            if (candidateTypes.length === 0) {
                // TODO find candidates based on most number of matching fields
                throw new Error(`No viable typenames found!${errorHint}`);
            } else if (candidateTypes.length === 1) {
                namedType = getNamedType(candidateTypes[0]);
            } else {
                // TODO filter down the list
                // a) Check the types of each field
                // b) Resolve unions
                throw new Error(`Cannot resolve the __typename between these possible types!

    ${_map(candidateTypes, candidateType => candidateType.inspect()).join('\n\t')}
${errorHint}`);
            }
        }
    }

    let type: GraphQLObjectType | GraphQLScalarType;
    if (isObjectType(namedType)) {
        type = namedType;
    } else if (isScalarType(namedType)) {
        return item;
    } else {
        if (isAbstractType(namedType)) {
            if (!namedType.resolveType) {
                // TODO inspect the options using the same field matching logic as is found in getCandidateTypes
                throw new Error(`No resolveType found for the interface ${namedType.name}`);
            }
            const typename = namedType.resolveType(item, null, {} as any, namedType);
            if (!typename) {
                throw new Error(`Unable to resolve type for abstract type ${namedType.name}`);
            }
            if (isPromise(typename)) {
                throw new Error(
                    `The resolveType for ${namedType.name} returns a promise but only synchronous resolvers are supported!`,
                );
            }
            if (isObjectType(typename)) {
                type = typename;
            } else {
                const resolvedType = schema.getType(typename);
                if (!resolvedType || !isObjectType(resolvedType)) {
                    throw new Error(
                        `The abstract type ${namedType.name} resolved to an unknown or misconfigured type ${typename}`,
                    );
                }
                type = resolvedType;
            }
        } else {
            throw new Error(
                `Unexpected type found ${namedType.name}! propertyOfParent : ${propertyOfParent}, parentTypeName: ${parentTypeName}`,
            );
        }
    }

    return _reduce(
        _keys(item),
        (newObj, propertyName) => {
            const child = item[propertyName];
            if (Array.isArray(child)) {
                newObj[propertyName] = _map(child, currentChild =>
                    addTypenames({
                        item: currentChild,
                        schema,
                        propertyOfParent: propertyName,
                        parentTypeName: type.name,
                    }),
                );
            } else {
                newObj[propertyName] = addTypenames({
                    item: child,
                    schema,
                    propertyOfParent: propertyName,
                    parentTypeName: type.name,
                });
            }
            return newObj;
        },
        {
            __typename: type.name,
        } as any,
    );
};
