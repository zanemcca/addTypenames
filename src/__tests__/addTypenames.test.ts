import { GraphQLSchema, buildSchema } from 'graphql';
import { DateTimeResolver, JSONObjectResolver } from 'graphql-scalars';
import { makeExecutableSchema } from '@graphql-tools/schema';
import _map from 'lodash/map';
import { readFileSync } from 'fs';

import { addTypenames } from '../addTypenames';

describe('addTypenames', () => {
    let schema: GraphQLSchema;
    let resolvers: any;
    let data: any;
    let typeDefs: string;
    beforeEach(() => {
        typeDefs = readFileSync(`${__dirname}/graphql.test.gql`, 'utf8');
        data = {
            chicken: {
                id: 1,
                name: 'Henrieta',
                cost: 45,
                feedRequirements: 45,
                eggOutput: 30,
            },
            cow: {
                id: 2,
                name: 'Moophy',
                cost: 1200,
                hayRequirements: 200,
                milkOutput: 500,
            },
            tractor: {
                id: 1,
                brand: 'John Deer',
                type: 'Tractor',
                purchasedOn: new Date(),
            },
            house: {
                id: 1,
                description: 'The house',
                cost: 400000,
            },
            barn: {
                id: 2,
                description: 'Barn',
                cost: 100000,
            },
        };
        resolvers = {
            DateTime: DateTimeResolver,
            JSONObject: JSONObjectResolver,
            Animal: {
                __resolveType: (animal: any) => {
                    if (animal.hayRequirements || animal.milkOutput) {
                        return 'Cow';
                    }
                    if (animal.eggOutput || animal.feedRequirements) {
                        return 'Chicken';
                    }
                    return null;
                },
            },
            Asset: {
                __resolveType: (asset: any) => {
                    if (asset.brand) {
                        return 'Equipment';
                    }
                    if (asset.description) {
                        return 'Building';
                    }
                    return null;
                },
            },
        };
        schema = makeExecutableSchema({
            typeDefs,
            resolvers,
        });
    });

    describe('trivial cases', () => {
        it('should return a string unmodified', () => {
            const input = 'hello world';
            expect(addTypenames({ item: input, schema })).toBe(input);
        });
        it('should return an int unmodified', () => {
            const input = 1234;
            expect(addTypenames({ item: input, schema })).toBe(input);
        });
        it('should return an float unmodified', () => {
            const input = 1234.567;
            expect(addTypenames({ item: input, schema })).toBe(input);
        });
    });

    it('should find the typenames of the children', async () => {
        const input = {
            __typename: 'Farm',
            buildings: [data.house, data.barn],
            equipment: [data.tractor],
        };
        const result = addTypenames({ item: input, schema });
        expect(result).toMatchObject({
            ...input,
            buildings: _map(input.buildings, building => ({
                ...building,
                __typename: 'Building',
            })),
            equipment: _map(input.equipment, equipment => ({
                ...equipment,
                __typename: 'Equipment',
            })),
        });
    });

    it('should resolve scalars', async () => {
        const input = {
            __typename: 'Farm',
            equipment: [
                {
                    ...data.tractor,
                    info: {
                        notes: 'Some notes',
                        vin: '123abc456',
                    },
                },
            ],
        };
        const result = addTypenames({ item: input, schema });
        expect(result).toMatchObject({
            ...input,
            equipment: _map(input.equipment, equipment => ({
                ...equipment,
                __typename: 'Equipment',
            })),
        });
    });

    it('should resolve interfaces', async () => {
        const input = {
            __typename: 'Farm',
            animals: [data.chicken, data.cow],
        };
        const result = addTypenames({ item: input, schema });
        expect(result).toMatchObject({
            ...input,
            animals: [
                {
                    ...data.chicken,
                    __typename: 'Chicken',
                },
                {
                    ...data.cow,
                    __typename: 'Cow',
                },
            ],
        });
    });

    it('should resolve unions', async () => {
        const input = {
            __typename: 'Farm',
            assets: [data.house, data.tractor, data.barn],
        };
        const result = addTypenames({ item: input, schema });
        expect(result).toMatchObject({
            ...input,
            assets: [
                {
                    ...data.house,
                    __typename: 'Building',
                },
                {
                    ...data.tractor,
                    __typename: 'Equipment',
                },
                {
                    ...data.barn,
                    __typename: 'Building',
                },
            ],
        });
    });

    describe('advanced usage', () => {
        describe('using propertyOfParent and parentTypeName', () => {
            it('should use the parentTypeName & propertyOfParent fields to resolve the ambiguous input', () => {
                const input = {
                    id: data.barn.id,
                };
                const result = addTypenames({
                    item: input,
                    schema,
                    parentTypeName: 'Farm',
                    propertyOfParent: 'equipment',
                });
                expect(result).toMatchObject({
                    ...input,
                    __typename: 'Equipment',
                });
            });

            describe('errors', () => {
                it('should complain about the propertyOfParent field being missing', () => {
                    const input = {
                        id: data.barn.id,
                    };
                    expect(() =>
                        addTypenames({ item: input, schema, parentTypeName: 'Farm' }),
                    ).toThrowErrorMatchingSnapshot();
                });
                it('should complain about the parentTypeName field being missing', () => {
                    const input = {
                        id: data.barn.id,
                    };
                    expect(() =>
                        addTypenames({ item: input, schema, propertyOfParent: 'equipment' }),
                    ).toThrowErrorMatchingSnapshot();
                });
            });
        });
    });

    describe('heuristic approach using getCandidateTypes', () => {
        it('should find the typename for a plain object when there is no ambiguity', async () => {
            const input = {
                id: 5,
                brand: 'International',
                type: 'Tractor',
            };
            const result = addTypenames({ item: input, schema });
            expect(result).toHaveProperty('__typename', 'Equipment');
        });

        describe('errors', () => {
            it('should complain about there being too many possible typename matches', async () => {
                const input = {
                    id: data.chicken.id,
                    name: data.chicken.name,
                };
                expect(() => addTypenames({ item: input, schema })).toThrowErrorMatchingSnapshot();
            });

            it('should complain about there being no typename matches', async () => {
                const input = {
                    ...data.chicken,
                    notAField: 'this is just some extra field that does not exist in the model',
                };
                expect(() => addTypenames({ item: input, schema })).toThrowErrorMatchingSnapshot();
            });
        });
    });

    describe('errors', () => {
        describe('async resolvers', () => {
            beforeEach(() => {
                const oldAnimalResolver = resolvers.Animal.__resolveType;
                resolvers = {
                    ...resolvers,
                    Animal: {
                        __resolveType: async (animal: any) => oldAnimalResolver(animal),
                    },
                };
                schema = makeExecutableSchema({
                    typeDefs,
                    resolvers,
                });
            });

            it('should complain about the Animal resolver using Promises', async () => {
                const input = {
                    __typename: 'Farm',
                    animals: [data.chicken],
                };
                expect(() => addTypenames({ item: input, schema })).toThrowErrorMatchingSnapshot();
            });
        });

        describe('missing resolvers', () => {
            beforeEach(() => {
                const oldAnimalResolver = resolvers.Animal.__resolveType;
                resolvers = {
                    ...resolvers,
                    Animal: {},
                };
                schema = makeExecutableSchema({
                    typeDefs,
                    resolvers,
                });
            });

            it('should complain about the Animal resolver not being present', async () => {
                const input = {
                    __typename: 'Farm',
                    animals: [data.chicken],
                };
                expect(() => addTypenames({ item: input, schema })).toThrowErrorMatchingSnapshot();
            });
        });
    });
});
