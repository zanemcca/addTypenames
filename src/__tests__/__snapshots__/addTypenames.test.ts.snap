// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTypenames advanced usage using propertyOfParent and parentTypeName errors should complain about the parentTypeName field being missing 1`] = `"Expected a parentTypeName to be given when propertyOfParent is given!"`;

exports[`addTypenames advanced usage using propertyOfParent and parentTypeName errors should complain about the propertyOfParent field being missing 1`] = `"Expected a propertyOfParent to be given when parentTypeName is given!"`;

exports[`addTypenames errors async resolvers should complain about the Animal resolver using Promises 1`] = `"The resolveType for Animal returns a promise but only synchronous resolvers are supported!"`;

exports[`addTypenames errors missing resolvers should complain about the Animal resolver not being present 1`] = `"No resolveType found for the interface Animal"`;

exports[`addTypenames heuristic approach using getCandidateTypes errors should complain about there being no typename matches 1`] = `
"No viable typenames found!

This is the relevant object given
{
	\\"id\\": 1,
	\\"name\\": \\"Henrieta\\",
	\\"cost\\": 45,
	\\"feedRequirements\\": 45,
	\\"eggOutput\\": 30,
	\\"notAField\\": \\"this is just some extra field that does not exist in the model\\"
}

Hint: Try adding a __typename on your initial input object or pass in a propertyOfParent & parentTypeName.
"
`;

exports[`addTypenames heuristic approach using getCandidateTypes errors should complain about there being too many possible typename matches 1`] = `
"Cannot resolve the __typename between these possible types!

    Animal
	Cow
	Chicken


This is the relevant object given
{
	\\"id\\": 1,
	\\"name\\": \\"Henrieta\\"
}

Hint: Try adding a __typename on your initial input object or pass in a propertyOfParent & parentTypeName.
"
`;
